Documentation Part I - General docu

JosePizarro3 opened the pull request:
@ndaelman-hu @JFRudzinski 

I created some docu page, and lay the structure for explaining this repo. Please, check it when you can. I will keep working on the `ModelSystem` part of the docu, as well as the other sections on "How to use the Simulation schema" and "The normalize function".

**Sorry for pushing very hard these things**, but I have the feeling we need to sprint a bit to have some stuff done before May-June. I don't want to impose anything, tho.

I want to use this docu page to open more the discussion with PIs in Area C (so that they can actually get involved and give feedback), as I am sure they will be happy to give feedback to the documentation page. I have the feeling this could evolved into an ontology very easily, but for this @ndaelman-hu can give his thoughts.

JosePizarro3 committed:
Added initial overview section and main structure

JosePizarro3 commented:
Ah, you can checkout locally and do `mkdocs serve` as usual, to visualize the MKDocs page.

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,101 @@
+<div class="click-zoom">
+    <label>
+        <input type="checkbox">
+        <img src="../assets/simulation_base.drawio.png" alt="Simulation base section diagram." width="80%" title="Click to zoom in">
```
What does the "s" in the diagrams following the type mean? I.e., `np.float64, s`? It is not immediately clear to me.

JosePizarro3 replied:
These are the units. Yeah, I agree it is pretty obscure. Should I remove them or come up with a more clear naming?

JosePizarro3 replied:
Added a note

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,101 @@
+    </label>
+</div>
+
+We use double inheritance from `EntryData` in order to populate the `data` section in the NOMAD archive. You can read more about the public normalize function function and its purpose in our base classes in the corresponding [documentation page](normalize.md). The private function `set_system_branch_depth()` is related with the `ModelSystem` base class, so we refer to its specific [documentation  ...
```
I would prefer using the name of the page for the link instead of the generic "documentation page":

"We use double inheritance from `EntryData` in order to populate the `data` section in the NOMAD archive. All of the base classes discussed here are subject to the [public normalize function](normalize.md) in NOMAD. The private function `set_system_branch_depth()` is related with the [ModelSystem base class](model_system/model_system.md)."

Check the statement about the normalize function, not sure if this is exactly what you meant, but just an example.

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,101 @@
+4. `Outputs`: contains all the output properties, as well as references to the `ModelSystem` and `ModelMethod` used to obtain such properties. It might also contain information which will populate `ModelSystem` (e.g., atomic occupations, atomic moments, crystal field energies, etc.).
+
+!!! note "Self-consistent steps, SinglePoint entries, and more complex workflows."
+    In NOMAD, we consider the minimal unit for storing the data in the archive (i.e., an *entry*) as any calculation containing all the self-consistent steps of itself. This is what we call, `SinglePoint`. Thus, we do not split each self-consistent step in its own entry in the NOMAD archive, but rather store them under the same entry in the archive. Any other complex calculation which combines se ...
```
Similar to above, I would replace "GitHub project" with something like "NOMAD simulation workflow schema"

JFRudzinski reviewed (approved):
@JosePizarro3 Looks like a great start, thanks a lot for taking the initiative and getting this started! Just a few minor points to start, but not important and anyway this will all develop as we go, I think you should add this or even any other additions as soon as you would like.

JosePizarro3 committed:
Changed way of rfcing to other docu in general.md

JosePizarro3 committed:
added docs for normalize function

JosePizarro3 committed:
Added Bernadette

Improved text

JosePizarro3 committed:
Improved plots quality

Added Bernadette email

JosePizarro3 committed:
Final improvement

JosePizarro3 commented:
@JFRudzinski @Bernadette-Mohr I finished this part of the documentation, leaving empty the specific sub-parts when talking about `ModelSystem`, `ModelMethod`, `Outputs` and on "How to use the schema", as we need more work before being able to write something.

There is also an open question at the end of "The normalize function" section.

Maybe you can drop a review? @Bernadette-Mohr I can help you visualize the MKDocs page locally so you can have a better idea of the docu page, let me know.

JosePizarro3 committed:
Delete comment in normalize.md

JosePizarro3 committed:
Improved normalize.md

JFRudzinski commented on docs/normalize.md:
```diff
@@ -0,0 +1,66 @@
+
+
+<!--
+JMP: I realized we have to make sure that a function has not been used already. Probably we can use the m_cache for this, but most importantly: does this make sense? I.e., do we want to overwrite running public functions when normalize() is ran or do we want to avoid this? Maybe someone used wrongly the function, or do we simply overlook this and ignore it?
```
I don't think I fully understand why we need to make sure that a public function is only used once. Are there not cases that you want to reuse a base class function multiple times? (I feel like I am missing something in the context here).

JosePizarro3 replied:
Check your last comment on this first review and my answer (I am deleting this section). I'll resolve this.

JFRudzinski commented on docs/contact.md:
```diff
@@ -4,17 +4,16 @@ NOMAD is an open source project that warmly welcomes community projects, contrib
+| Dr. Nathan Daelman | [nathan.daelman@physik.hu-berlin.de](mailto:nathan.daelman@physik.hu-berlin.de) | DFT | [@ndaelman-hu](https://github.com/ndaelman-hu) |
+| Dr. Bernadette Mohr | [mohrbern@physik.hu-berlin.de](mailto:mohrbern@physik.hu-berlin.de) | MD, FF | [@Bernadette-Mohr](https://github.com/Bernadette-Mohr) |
+| Dr. José M. Pizarro | [jose.pizarro@physik.hu-berlin.de](mailto:jose.pizarro@physik.hu-berlin.de) | GW, DMFT, BSE | [@JosePizarro3](https://github.com/JosePizarro3) |
+| Dr. Joseph F. Rudzinski (**Coordinator**) | [joseph.rudzinski@physik.hu-berlin.de](mailto:joseph.rudzinski@physik.hu-berlin.de) | MD, FF | [@JFRudzinski](https://github.com/JFRudzinski) |
```
Can you change my topics to something like "General" (maybe you have an idea for better wording). I would like to give the impression that I am there for overarching issues / general contact.

JFRudzinski commented on docs/contact.md:
```diff
@@ -4,17 +4,16 @@ NOMAD is an open source project that warmly welcomes community projects, contrib
-- Open an issue in the [general NOMAD Github project](https://github.com/nomad-coe/nomad), or in one of the [sub-projects](https://github.com/nomad-coe/nomad/tree/develop/dependencies/parsers) related with specific parsers. Our Github profile tags are [@ndaelman-hu](https://github.com/ndaelman-hu), [@ladinesa](https://github.com/ladinesa), [@JosePizarro3](https://github.com/JosePizarro3), and [@J ...
-- Write us in the [NOMAD MatSci forum](https://matsci.org/c/nomad/32). Our tags there are @NateD, @ladinesa, @JosePizarro, and @JFRudzinski.
-- Send an email to [support@nomad-lab.eu](mailto:support@nomad-lab.eu). Please, add in the subject "ATTN - Area C".
+- Open an [**issue**](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/issues) in the [Github project](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/), and tag any of us.
```
Should we suggest that they join the Discord? I would opt for this before sending an email to support, possibly as a first point of contact. In that case, maybe add our Discord handles to the table.

JosePizarro3 replied:
Done, tho Discord recognizes our names, so I wouldn't add our handles

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,106 @@
+# `Simulation` base class
+
+We defined a base class named `Simulation`. You can find its Python schema definition in [src/nomad_simulations/general.py](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/blob/develop/src/nomad_simulations/general.py). In NOMAD, this section will
```
I suggest starting with something like: "In NOMAD, all simulation metadata is defined within the `Simulation` base class." 

Then, I would remove "In NOMAD" in the next few sentences.

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,106 @@
+# `Simulation` base class
+
+We defined a base class named `Simulation`. You can find its Python schema definition in [src/nomad_simulations/general.py](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/blob/develop/src/nomad_simulations/general.py). In NOMAD, this section will
```
base class == base section? And if so, is this characterization of simulation as a base class correct? Based on the link below, this implies that Simulation will be found under `data.metainfo.basesections` (People are using "base class" now pretty loosly, so I just want to hear your interpretation of what this means).

JosePizarro3 replied:
Ok, let me rephrase these 3 initial paragraphs and fix the first image.

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,106 @@
+We defined a base class named `Simulation`. You can find its Python schema definition in [src/nomad_simulations/general.py](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/blob/develop/src/nomad_simulations/general.py). In NOMAD, this section will
+appear under the `data` section for the _archive_ metadata structure of each _entry_.
+
+In NOMAD, a set of [base sections](https://nomad-lab.eu/prod/v1/staging/docs/howto/customization/base_sections.html) derived from the [Basic Formal Ontology (BFO)](https://basic-formal-ontology.org/) are defined. We used them to define `Simulation` as an [`Activity`](http://purl.obolibrary.org/obo/BFO_0000015). The UML diagram is:
```
I think I would opt to rephrase this slightly, but it depends a little on your answer to my previous question. So, I put this placeholder and will come back later...

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,106 @@
+    </label>
+</div>
+
+In fact, `Simulation` inherits from a further abstract concept, `BaseSimulation`. This class contains the general information about the `Program` used, as well as general times of the simulation, e.g., the datetime at which it started (`datetime`) and ended (`datetime_end`). The detailed UML diagram of quantities and functions defined for `Simulation` is thus:
```
concept or class?

JosePizarro3 replied:
Class is more proper, yes. But let me rephrase a bit these 3 first paragraphs.

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,106 @@
+4. `Outputs`: contains all the output properties, as well as references to the `ModelSystem` used to obtain such properties. It might also contain information which will populate `ModelSystem` (e.g., atomic occupations, atomic moments, crystal field energies, etc.).
+
+!!! note "Self-consistent steps, SinglePoint entries, and more complex workflows."
+    In NOMAD, we consider the minimal unit for storing the data in the archive (i.e., an *entry*) as any calculation containing all the self-consistent steps of itself. This is what we call, **`SinglePoint`**. Thus, we do not split each self-consistent step in its own individual entry in the NOMAD archive, but rather store them under the same entry in the archive. Any other complex calculation wh ...
```
This is a bit confusing because it is conflating the definition of the workflow section with the contents of an archive. But in practice, these 2 things are considered separately and depend on the situation. In particular, you may have a workflow whose steps are all contained within a single entry. But a singlepoint self-consistent field calculation does not contain a workflow section.

**First** attempt at rewrite:
"""
The minimal unit for storing data in the NOMAD archive is an [*Entry*](<link to gen docs glossary?>). In the context of simulation data, an entry may contain data from a calculation on an individual system configuration (e.g., a single-point DFT calculation) using **only** the above-mentioned sections of the `Simulation` class. Information from self-consistent iterations to converge properties for this configuration are also contained within these sections.

More complex calculations that involve multiple configurations require the definition of a [**Workflow**](<?>) section within the archive. Depending on the situation, the information from individual workflow steps may be stored within a single or multiple entries. For example, for efficiency, the data from workflows involving a large amount of configurations, e.g., molecular dynamics trajectories, are stored within a single entry. Other standard workflows store the single-point data in separate entries, e.g.,  a `GW` calculation is usually composed of a `DFT SinglePoint` entry and a `GW SinglePoint` entry. Higher-level workflows, which simply connect a series of standard or custom workflows, are typically stored as a separate entry. You can check the [NOMAD simulations workflow schema](https://github.com/nomad-coe/nomad-schema-plugin-simulation-workflow) for more information.
"""

JosePizarro3 replied:
Yeah, I like it, thanks a lot, I was struggling finding how to explain it easily. I incorporate it to the note.

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,106 @@
+!!! note "Self-consistent steps, SinglePoint entries, and more complex workflows."
+    In NOMAD, we consider the minimal unit for storing the data in the archive (i.e., an *entry*) as any calculation containing all the self-consistent steps of itself. This is what we call, **`SinglePoint`**. Thus, we do not split each self-consistent step in its own individual entry in the NOMAD archive, but rather store them under the same entry in the archive. Any other complex calculation wh ...
+
+The simplified schematics for a `Simulation` data section will then be (note that the arrows here are a simple way of visually defining _inputs_ and _outputs_):
```
The following schematic represents a simplified representation of the `Simulation` section (note...):

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,106 @@
+
+### `Program` {#program}
+
+The `Program` base class section contains all the information about the program / software / code used to perform the simulation. We considered to be a [`(Continuant) Entity`](http://purl.obolibrary.org/obo/BFO_0000002). It is contained within `BaseSimulation` as a sub-section.. The detailed UML diagram is:
```
"We consider it to be a ..., contained as a sub-section within `BaseSimulation`"

JFRudzinski commented on docs/general.md:
```diff
@@ -0,0 +1,106 @@
+</div>
+
+
+When [writing a parser](https://nomad-lab.eu/prod/v1/staging/docs/howto/customization/parsers.html), we recommend to start by instantiating the `Program` section and populating its quantities to get acquaintant and learn how to use the NOMAD parsing infrastructure.
```
"...its quantities, in order to get acquainted with the NOMAD parsing infrastructure" (it's a bit redundant with both acquainted and learn)

JFRudzinski commented on docs/general.md:
````diff
@@ -0,0 +1,106 @@
+...
+```
+
+Then, we can parse the program `name` and `version` by matching the texts (see, e.g., [Wikipedia page for Regular expressions, also called _regex_](https://en.wikipedia.org/wiki/Regular_expression)):
````
~~Then,~~ We can...

JFRudzinski commented on docs/index.md:
```diff
@@ -3,17 +3,19 @@
+When designing the sections, we follow [SOLID principles](https://www.geeksforgeeks.org/solid-principle-in-programming-understand-with-real-life-examples/) for object-oriented programming. And throughout this documentation, we will use [UML diagrams](https://en.wikipedia.org/wiki/Class_diagram), both in a simplified and in a detailed manner, to draw the schemas relationships.
 
 
 !!! info "Main contributors"
```
This seems a bit unnecessary with the contact page. Maybe you can just add a link or button to there?

JFRudzinski commented on docs/normalize.md:
```diff
@@ -0,0 +1,66 @@
+# The `normalize()` function
+
+Each base class defined using the NOMAD schema has a set of public functions which can be used at any moment when reading and parsing files in NOMAD (see their corresponding documentation pages). The only special situation happens for the `normalize(archive, logger)` function.
```
I wouldn't say "see their corresponding..." unless you are going to link something.

"The only special situation..." --> "The `normalize()` function is a special case of such a function, which warrants an in-depth description"

JFRudzinski commented on docs/normalize.md:
```diff
@@ -0,0 +1,66 @@
+
+Each base class defined using the NOMAD schema has a set of public functions which can be used at any moment when reading and parsing files in NOMAD (see their corresponding documentation pages). The only special situation happens for the `normalize(archive, logger)` function.
+
+This function is ran within the NOMAD infrastructure by the [`MetainfoNormalizer`](https://github.com/nomad-coe/nomad/blob/develop/nomad/normalizing/metainfo.py) in the following order:
```
"is run"

JFRudzinski commented on docs/normalize.md:
```diff
@@ -0,0 +1,66 @@
+
+This function is ran within the NOMAD infrastructure by the [`MetainfoNormalizer`](https://github.com/nomad-coe/nomad/blob/develop/nomad/normalizing/metainfo.py) in the following order:
+
+1. The deepest child class `normalize()` function is ran before their parents' `normalize()` function.
```
"A child class' ... is run before their/its parents' ..."

JFRudzinski commented on docs/normalize.md:
```diff
@@ -0,0 +1,66 @@
+This function is ran within the NOMAD infrastructure by the [`MetainfoNormalizer`](https://github.com/nomad-coe/nomad/blob/develop/nomad/normalizing/metainfo.py) in the following order:
+
+1. The deepest child class `normalize()` function is ran before their parents' `normalize()` function.
+2. For sibling sections, the `normalize()` function is executed from the smaller to the larger `normalizer_level` attribute. If `normalizer_level` is not set up, the order is established by the attributes definition order in the parent class.
```
"not set up" --> "not set"?

Also, what if they are set to be the same, I guess then it defaults again to the order in the parent class?

JosePizarro3 replied:
> Also, what if they are set to be the same, I guess then it defaults again to the order in the parent class?

Yes, I added "If `normalizer_level` is not set or if they are the same for two different sections, the order..."

JFRudzinski commented on docs/normalize.md:
````diff
@@ -0,0 +1,66 @@
+        # Some operations here or before `super().normalize(archive, logger)`
+```
+
+Now, `MetainfoNormalizer` will be ran on the `ParentSection`. Applying **rule 1**, the `normalize()` functions of the `ParentSection`'s childs are executed first. The order of these functions is established by **rule 2** with the `normalizer_level` atrribute, i.e., all the `Section2` (note that `sub_section_2` is a list of sections) `normalize()` functions are ran first, then `Section1.normalize( ...
````
"will be run" and "functions are run first" (The 2nd one is tricky, but I think that is correct)

JFRudzinski commented on docs/normalize.md:
```diff
@@ -0,0 +1,66 @@
+2. `Section1.normalize()`
+3. `ParentSection.normalize()`
+
+In case we do not assign a value to `Section1.normalizer_level` and `Section2.normalizer_level`, all `Section1.normalize()` will run first and then `Section2.normalize()` due to the order of `SubSection` attributes in `ParentSection`. Thus the order will be in this case:
```
"~~all~~ `Section1.normalize()` will run ~~and then~~ before `Section2.normalize`, due to the ..."

JFRudzinski commented on docs/normalize.md:
```diff
@@ -0,0 +1,66 @@
+2. `Section2.normalize()`
+3. `ParentSection.normalize()`
+
+By checking on the `normalize()` functions and **rule 3**, we can establish whether `ArchiveSection.normalize()` will be ran or not. In `Section1.normalize()`, it will not, while in the other sections, `Section2` and `ParentSection`, it will.
```
"will be run"

JosePizarro3 replied:
yeah, at some point you would imagine that I know that, but it is obvious that it is not the case 🤣

JFRudzinski commented on docs/normalize.md:
```diff
@@ -0,0 +1,66 @@
+
+### Avoiding double execution of public functions
+
+When the `MetainfoNormalize` is ran, and the `normalize()` functions are executed, these usually have calls for calling private and public functions in the base classes which might have been used already during parsing the files. However, we have checks to ensure that this will not happen. As an user, you can decide whether running then in your parsers for helping you during the process, or leave ...
```
"is run"

JFRudzinski replied:
"as a user"

JFRudzinski replied:
"whether running **them**"

JFRudzinski replied:
"these usually call private and..."?

JFRudzinski replied:
"As a user, you can decide whether to run them directly in within your parser or to leave it up to the `normalize()` function"?

Something like this, but I don't yet understand exactly what the point is...

JosePizarro3 replied:
Yes, I was thinking that the current implementation would be more optimized if we have someway of controlling if a function is executed or not to avoid double executions in the `normalize` function, but I think we don't want that, we anyways want the normalization to run.

JFRudzinski reviewed (commented):
I'll come back to expand upon some things after you respond to a few of my comments

JosePizarro3 committed:
Fix initial paragraphs in general.md

Improved contact and index

ndaelman-hu commented on docs/contact.md:
```diff
@@ -4,17 +4,16 @@ NOMAD is an open source project that warmly welcomes community projects, contrib
-    | Dr. Joseph F. Rudzinski (**Coordinator**) | [joseph.rudzinski@physik.hu-berlin.de](mailto:joseph.rudzinski@physik.hu-berlin.de)   | MD, parsers, workflows, normalizers                          |
+| Name | E-mail     | Topics | Github profiles |
+|------|------------|--------|-----------------|
+| Dr. Nathan Daelman | [nathan.daelman@physik.hu-berlin.de](mailto:nathan.daelman@physik.hu-berlin.de) | DFT | [@ndaelman-hu](https://github.com/ndaelman-hu) |
```
Electronic structure, basis sets?

JosePizarro3 replied:
I think DFT is pretty good. No need to specify to such level of detail in here.

ndaelman-hu replied:
People should be able to contact someone regarding the basis sets, band structure and DOS. Who isn't made apparent right now. Add it to my info, pls.

ndaelman-hu commented on docs/contact.md:
Typically, contact information, goes into the README, which is shown at the front of the Github page.

JosePizarro3 replied:
Ok, I added it, but I also kept it in the documentation page.

ndaelman-hu commented:
> I want to use this docu page to open more the discussion with PIs in Area C (so that they can actually get involved and give feedback), as I am sure they will be happy to give feedback to the documentation page. I have the feeling this could evolved into an ontology very easily, but for this @ndaelman-hu can give his thoughts.

Yes, this would be very helpful for understanding some of the design choices. Sometimes, it's hard to read these from the code alone.

Regarding the ontology, going off your diagrams, most of them are UML class diagrams, which already have a well-defined semantics. The exception here is the program scheme, but there the arrows are unspecified.

Btw, I wouldn't target top-level ontologies, as designing them probably goes over our heads, and their space is already saturated.
I think the field is more so in dire need of domain-specific ontologies. Our schema could form the basis of such an ontology, yes. In the end, this is just a well-defined nomenclature. Getting the community to adopt it is the tricky part. If they did, it would help tremendously with parsing, though.

ndaelman-hu commented on docs/assets/program.png:
I would state the legend as:
- "inherits from"
- "is contained in" + specify the meaning of the label

JosePizarro3 replied:
I think the labels are understandable enough. I will add the description of the of the label when containment, tho.

ndaelman-hu commented on docs/assets/simulation_base.png:
Rather than inheritance trees, I'd prefer seeing the interaction modes between objects.

JosePizarro3 replied:
Not sure what you mean.

ndaelman-hu replied:
_class inheritance_ is used as a way for extending user interfaces and functionality. I don't get from this figure what or why it is being extended. the figure gives some overview, but it'll get old soon, and I can just trace the inheritance with the IDE.

As Alan Kay put it, the essence of OOP is _the concept of message passing_ (I guess that in physics, we'd rather say _how 2 objects interact_). Class UML diagrams try to convey this too.

JosePizarro3 replied:
I still don't understand what you are trying to say. I have no idea of who Alan Kay is.

This figure is kind of a simplified diagram of the inheritance tree. At this point, we don't want to show all the set of quantities, defs, and methods.

ndaelman-hu commented on docs/assets/simulation_composition.png:
label the arrows

JosePizarro3 replied:
What do you mean?

ndaelman-hu replied:
The arrows denote some kind of connection. Which is ..?
I take it they all have the same semantics, though, since the same arrow format is used.

You can also add it as a legend.

JosePizarro3 replied:
This is a simple input/output arrow, it is explained in the text.

ndaelman-hu reviewed (commented):
short feedback for now

JosePizarro3 committed:
Change classes to sections

Added contact to README

Deleted outdated packages in MKDocs

Added feedback

Fix simulation_base.png

JosePizarro3 commented:
Ok, I think I can merge this and keep working on a second and third parts. 

Please, do a quick check so everything looks reasonable :)

JosePizarro3 commented:
@JFRudzinski when you can, do you mind checking again the initial paragraphs in `general.md`? Let me know if now is more clear.

JFRudzinski commented:
> @JFRudzinski when you can, do you mind checking again the initial paragraphs in `general.md`? Let me know if now is more clear.

Yeah, this looks better :+1:

JosePizarro3 commented:
Cool, I merge then.

Thanks for the reviews, @ndaelman-hu @JFRudzinski ! 👍🏻

JosePizarro3 committed:
Final fixes
